### ISSUE TITLE
Rockchip NPU support in whisper.cpp

### ISSUE THREAD

**Issue by bobqianic** [ggerganov/whisper.cpp#1557]
Christmas is coming soon, and I want to take some time to research something interesting, such as edge low-power inference. Although current whisper.cpp can run on [Raspberry Pi](https://github.com/ggerganov/whisper.cpp/discussions/166), the inference performance cannot achieve real-time transcription. Fortunately, there are now some development boards that use processors with NPUs, which can be used to achieve real-time transcription of large models. My primary goal is to first support RK3566 and RK3588.

### Roadmap:
- [ ] MatMul offloading
- [ ] Conv-Gelu offloading
- [ ] LayerNorm offloading
 ...

### Reference:
https://github.com/rockchip-linux/rknpu2

**Comment by ggerganov** [comment]
Would be great if we can find a way to utilize the NPUs! Keep us in the loop!

**Comment by Leeviber** [comment]
I tried converting the whisper encode model to rknpu format(.rknn), it succeeded but the estimated runtime is quite slow, even lower than running on CPU. I think the NPU is not full support transformer, some operators are still running on the CPU.

**Comment by RoboMagus** [comment]
Some interesting development was done here: https://github.com/usefulsensors/useful-transformers.

However not everything runs on the NPU and I've personally had mixed success on running non English models.

**Comment by bobqianic** [comment]
> Some interesting development was done here: https://github.com/usefulsensors/useful-transformers.

Yes, I've seen that. But I'm looking to enhance the ggml tensor library by adding some operators. This way, not only will whisper.cpp be able to utilize the NPU, but other ggml examples like llama.cpp as well. I've ordered an OrangePi 5 Plus with 32GiB RAM from Aliexpress, which is still in transit :  )

> However not everything runs on the NPU and I've personally had mixed success on running non English models.

Hopefully, we'll be able to run all models, regardless of their size, and whether they are English-only or support multiple languages.

**Comment by bobqianic** [comment]
The most challenging aspect I've encountered thus far is finding an appropriate driver for the RK3588 & RK3566 NPU. Most Linux distributions don't include an NPU driver, with this one being the notable exception. 

https://github.com/unifreq/linux-5.10.y-rk35xx/tree/main/drivers/rknpu

**Comment by bobqianic** [comment]
> I tried converting the whisper encode model to rknpu format(.rknn), it successed but the estimated runtime is quite slow, even lower than running on CPU. I think the NPU is not full support transformer, some operators are still running on the CPU.

You're right. From my experiments, it seems the NPU on the RK3588 is only effective for 3x3 convolutions. Unfortunately, its GEMM performance is **_quite poor_**. Despite being equipped with a `3x2 TOPs` NPU, each unit only delivers about `10 GFLOPs` for FP16 GEMM or `20 GFLOPs` for INT8 GEMM. It's quite a letdown. I regret to share such disappointing news during the holiday.

![image](https://github.com/ggerganov/whisper.cpp/assets/129547291/e406805b-8b30-493f-b12f-0d6241f0ab2b)

**Comment by bobqianic** [comment]
I discovered that someone else did the exact same thing but didn't find success. @ggerganov 

The challenge with the Rockchip NPU stems from its peculiar input and output dimensions. To attain maximum speed, it's necessary to transform a 2D matrix into a particular dimension. If you don't do this, the driver will take over, **but it operates much slower**. After processing, you need to convert the result back to its original dimension. This process is quite inefficient, and I'm sharing this to prevent others from spending unnecessary time trying to implement it.

With the RK3588, when you're working with a matrix A of size `(N, K)` and a matrix B of size `(K, M)`, you'll need to reshape matrix A to the new dimensions of `(K/8, N, 8)`. Similarly, reshape matrix B to `(M/16, K/32, 16, 32)`. After these transformations, the resulting output matrix C will have the dimensions of `(N/4, M, 4)`, instead of the expected `(N, M)`.

Links: 
https://clehaxze.tw/gemlog/2023/12-17-update-on-ggml-rknpu2-backend-and-rknpu2-1_6_0.gmi
https://github.com/marty1885/llama.cpp/tree/rknpu2-backend

**Comment by Leeviber** [comment]
Which RKNPU kernel driver version you using for above test ? 
From their [newest changelog,](https://github.com/airockchip/rknn-toolkit2?tab=readme-ov-file#160) the rknpu1.6 have better support for transformer ops and multi-dimension matmul.

**Comment by ggerganov** [comment]
@bobqianic Thank you for the updates! The work in [marty1885/llama.cpp@rknpu2-backend](https://github.com/marty1885/llama.cpp/tree/rknpu2-backend?rgh-link-date=2023-12-24T21%3A39%3A39Z) is interesting and I will be following with the progress

**Comment by marty1885** [comment]
For reference. People have worked around the matrix reordering specifically for Whisper by abstracting the entire thing around that fact.

useful-transformers is a very successful implementation. https://github.com/usefulsensors/useful-transformers

**Comment by Lhemamou** [comment]
Hey :) as raspberry is launching a new TPU hat https://www.raspberrypi.com/products/ai-kit/ I am reopening the topic. Do you have by chance any news or ways to begin enhance performance thanks to this hat ? I guess it would be easier than coral as we don't need to jailbreak it.

**Comment by marty1885** [comment]
@Lhemamou I actually talked to Halio about this during Computex. Long story short. No unless someone wants to form a company and sign an NDA to gain low level access.

**Comment by solarsamuel** [comment]
@marty1885 I have a company and I'd be open to signing a NDA as long as it looks reasonable, but before I go too far, my main concern is in regard to hardware. 

Does anyone know what the hailo hardware limit is in regard to model size? Feel free to send links. 

For example, the Google Coral TPU stick ASIC has 8MB of SRAM built into the chip. Something like 1.5MB of overhead is used, so a model can only be 6.5MB max. https://coral.ai/docs/edgetpu/compiler/#parameter-data-caching

For the Google Coral TPU the whisper tiny model is too big, even the 4 bit quantized version of the tiny model is around 24MB. 

tiny | 75 MiB disk | ~273 MB Mem

I'm assuming the Hailo chip does the matrix multiply internally and the results are stored in a pipeline in internal SRAM, but I could be wrong.

**Comment by marty1885** [comment]
@solarsamuel I can't tell without knowing NDAed information. From what I gathered from their sales and engineers.

1. The Hailo 8 can fit YOLOv5s and a modified version of YOLOv5m
2. If their compiler cannot fit the model onto the chip. They can split the model and switch the weights on the fly
  * But sure, there will be performance impact. Limited by PCIe bandwidth
  * Or split the model across multiple chips
3. Hailo-10H has DRAM. You can put large models there. It eliminates PCIe transfer, now the bottleneck is DRAM bandwidth
4. whisper.cpp requires low level access to the accelerator. It needs to be able to command the accelerator to do matmul directly. A compiler layer is useless in this case. If you want to sign an NDA, you need to check that you also get that level of access.

**Comment by solarsamuel** [comment]
@marty1885 I can reach out. Who would be a good person to contact? I'm definitely not making any guarantees any of this will work out.

**Comment by marty1885** [comment]
@solarsamuel Sorry for the late reply. I got caught in some personal issues. Let's not misuse the issue tracker and talk through email? You can find mine on my website via the link on my GitHub profile. 

Your GH profile links to a company and I'm not sure if that's the one you want to use for discussion.

**Comment by jenskastensson** [comment]
@bobqianic - Would you benefit from having a driver package for Mali GPU kernel drivers on RK3588 (specifically for Debian Bullseye)? Let me know if this is something that would improve inference performance  !
```
/opt/libmali# dmesg | grep mali
[    3.406183] mali fb000000.gpu: Kernel DDK version g21p0-01eac0
[    3.406423] mali fb000000.gpu: Looking up mali-supply from device tree
[    3.406569] mali fb000000.gpu: Looking up mem-supply from device tree
[    3.406849] mali fb000000.gpu: Looking up mali-supply from device tree
[    3.407154] mali fb000000.gpu: bin=0
[    3.407392] mali fb000000.gpu: leakage=16
[    3.407516] mali fb000000.gpu: Looking up mali-supply from device tree
[    3.407547] debugfs: Directory 'fb000000.gpu-mali' with parent 'vdd_gpu_s0' already present!
[    3.408956] mali fb000000.gpu: pvtm=865
[    3.409284] mali fb000000.gpu: pvtm-volt-sel=3
[    3.409338] mali fb000000.gpu: Looking up mali-supply from device tree
[    3.409358] debugfs: Directory 'fb000000.gpu-mali' with parent 'vdd_gpu_s0' already present!
[    3.409375] mali fb000000.gpu: Looking up mem-supply from device tree
[    3.410738] mali fb000000.gpu: avs=0
[    3.410849] W : [File] : drivers/gpu/arm/bifrost/platform/rk/mali_kbase_config_rk.c; [Line] : 144; [Func] : kbase_platform_rk_init(); power-off-delay-ms not available.
[    3.411650] mali fb000000.gpu: Register LUT 000a0800 initialized for GPU arch 0x000a0806
[    3.411678] mali fb000000.gpu: r0p0 status 5 not found in HW issues table;
[    3.411689] mali fb000000.gpu: falling back to closest match: r0p0 status 0
[    3.411698] mali fb000000.gpu: Execution proceeding normally with fallback match
[    3.411706] mali fb000000.gpu: GPU identified as 0x7 arch 10.8.6 r0p0 status 0
[    3.411810] mali fb000000.gpu: No priority control manager is configured
[    3.411819] mali fb000000.gpu: Large page allocation set to false after hardware feature check
[    3.412265] mali fb000000.gpu: No memory group manager is configured
[    3.412300] mali fb000000.gpu: Protected memory allocator not available
[    3.413941] mali fb000000.gpu: EM: OPP:600000 is inefficient
[    3.413953] mali fb000000.gpu: EM: OPP:500000 is inefficient
[    3.413961] mali fb000000.gpu: EM: OPP:400000 is inefficient
[    3.413968] mali fb000000.gpu: EM: OPP:300000 is inefficient
[    3.414342] mali fb000000.gpu: EM: created perf domain
[    3.414989] mali fb000000.gpu: l=10000 h=85000 hyst=5000 l_limit=0 h_limit=800000000 h_table=0
[    3.415901] mali fb000000.gpu: * MALI kbase_mmap_min_addr compiled to CONFIG_DEFAULT_MMAP_MIN_ADDR, no runtime update possible! *
[    3.415920] mali fb000000.gpu: Probed as mali0
[    7.404098] mali fb000000.gpu: Loading Mali firmware 0x1010000
[    7.406892] mali fb000000.gpu: Mali firmware git_sha: ee476db42870778306fa8d559a605a73f13e455c
```

**Comment by kcoul** [comment]
I'm becoming increasingly interested in getting this to work. Hailo-10H could potentially accelerate both whisper.cpp and llama.cpp with a small enough model. I am always getting annoyed trying to do things in Python (which Hailo already supports, but dev env is quite flaky in my experience), so this would be a good step forward.
